mono/utils/mono-semaphore.c

```diff
@@ -17,67 +17,78 @@
 #include <unistd.h>
 #endif
 
-#if (defined (HAVE_SEMAPHORE_H) || defined (USE_MACH_SEMA)) && !defined(HOST_WIN32)
-/* sem_* or semaphore_* functions in use */
-#  ifdef USE_MACH_SEMA
-#    define TIMESPEC mach_timespec_t
-#    define WAIT_BLOCK(a,b) semaphore_timedwait (*(a), *(b))
-#  elif defined(__native_client__) && defined(USE_NEWLIB)
-#    define TIMESPEC struct timespec
-#    define WAIT_BLOCK(a, b) sem_trywait(a)
-#  else
-#    define TIMESPEC struct timespec
-#    define WAIT_BLOCK(a,b) sem_timedwait (a, b)
-#  endif
+#include "mono-semaphore.h"
+#include "mono-threads-coop.h"
 
 #ifndef NSEC_PER_SEC
-#define NSEC_PER_SEC 1000000000
+#define NSEC_PER_SEC 1000 * 1000 * 1000
 #endif
 
+#if defined(USE_MACH_SEMA)
+
+int
+mono_sem_init (MonoSemType *sem, int value)
+{
+	return semaphore_create (current_task (), sem, SYNC_POLICY_FIFO, value) != KERN_SUCCESS ? -1 : 0;
+}
+
+int
+mono_sem_destroy (MonoSemType *sem)
+{
+	return semaphore_destroy (current_task (), *sem) != KERN_SUCCESS ? -1 : 0;
+}
+
+int
+mono_sem_wait (MonoSemType *sem, gboolean alertable)
+{
+	int res;
+
+	MONO_TRY_BLOCKING;
+
+retry:
+	res = semaphore_wait (*sem);
+	g_assert (res != KERN_INVALID_ARGUMENT);
+
+	if (res == KERN_ABORTED && !alertable)
+		goto retry;
+
+	MONO_FINISH_TRY_BLOCKING;
+
+	return res != KERN_SUCCESS ? -1 : 0;
+}
+
 int
 mono_sem_timedwait (MonoSemType *sem, guint32 timeout_ms, gboolean alertable)
 {
-	TIMESPEC ts, copy;
-	struct timeval t;
+	mach_timespec_t ts, copy;
+	struct timeval start, current;
 	int res = 0;
 
-#ifndef USE_MACH_SEMA
-	if (timeout_ms == 0)
-		return sem_trywait (sem);
-#endif
 	if (timeout_ms == (guint32) 0xFFFFFFFF)
 		return mono_sem_wait (sem, alertable);
 
-#ifdef USE_MACH_SEMA
-	memset (&t, 0, sizeof (t));
-#else
-	gettimeofday (&t, NULL);
-#endif
-	ts.tv_sec = timeout_ms / 1000 + t.tv_sec;
-	ts.tv_nsec = (timeout_ms % 1000) * 1000000 + t.tv_usec * 1000;
+	ts.tv_sec = timeout_ms / 1000;
```

Comment:
couldn't you move `MONO_TRY_BLOCKING` up here instead of line 80? You could block slightly longer.